keps/sig-auth/20191101-add-request-id-to-k8s-logs.md

```diff
@@ -0,0 +1,322 @@
+ - Case1. The function which contains Tracing codes by `Distributed Tracing` KEP
+   - We just add Request-ID codes (Request-ID codes don't interfere Tracing codes).
+ - Case2. The function which does not contain Tracing codes by `Distributed Tracing` KEP
+   - We add both of Tracing codes and Request-id codes (Request-ID codes don't interfere Tracing codes).
+
+#### Design of Export Request-ID
+
+We get Trace-ID by using `Distributed Tracing` feature in each Kubernetes function that we add Tracing codes. And then, we add Trace-ID information into **each klog call** as Request-ID. Note that we don't associate Request-ID to all of klog calls. Our target is important operations such as `kubectl create/delete/etc.`, and our target klog calls are the only klogs which is called via such important operations. Request-ID feature does not change existing klog function/method, but changes each klog calls and their log format. Currently, there is [Structured logging](https://github.com/serathius/enhancements/blob/structured-logging/keps/sig-instrumentation/20191115-structured-logging.md) KEP, and this KEP also change specific klog calls. We will merge Request-ID feature after Structured logging KEP is merged.
+
+**Examples (We quote some parts from `Structured logging` KEP)**
+
+Source Code
+
+Original
+```go
+klog.Infof("Updated pod %s status to ready", pod.name)
+```
+
+Structured format
+```go
+klog.InfoS("Pod status updated", "pod", pod, "status", "ready")
+```
+Request-ID with structured format
+```go
+klog.InfoS("Pod status updated", "pod", pod, "status", "ready", "Request-ID", trace-id)
+```
+
+Expected Log
+
+Structured format
+```json
+{
+   "ts": 1580306777.04728,
+   "v": 4,
+   "msg": "Pod status updated",
+   "pod":{
+      "name": "nginx-1",
+      "namespace": "default"
+   },
+   "status": "ready"
+}
+
+Request-ID with structured format
+```json
+{
+   "ts": 1580306777.04728,
+   "v": 4,
+   "msg": "Pod status updated",
+   "pod":{
+      "name": "nginx-1",
+      "namespace": "default"
+   },
+   "status": "ready"
+   "request-id": 5acf2a4d258157e06402fb734186b684
+}
+```
+#### Design overview of Control Request-ID
+
+We should control Request-ID feature to avoid an impact to existing users who are retrieving logs and analyzing with existing log format. So we introduce `--request-id` parameter which enables/disables Request-ID feature. We also manage the range of operations which are added Request-ID. The effect of each parameter of `--request-id` is as follows.
+
+| parameter | efficient |
+| ------ | ------ |
+| --request-id=0 | Request-ID feature is disabled (Default) |
+| --request-id=1 | Request-ID feature is enabled, and Request-ID is added to klogs related to the `Alpha` target operations | 
+| --request-id=2 | Request-ID feature is enabled, and Request-ID is added to klogs related to the `Alpha and Beta` target operations | 
```

Comment:
Interesting.  This is the equivalent of verbosity for adding this info to logs.  How do you plan to propagate the request id value here to controllers?  Will it also have to be stored in an annotation and propagated the same way the SpanContext is for tracing?

Reply:
I am considering the method, but I don't have the best idea now.
Currently, I think we don't use the same way the SpanContext. There is an idea that we add parameter into component-base. This is the same idea as structured logging(--logging-format). However, I think this requires restart of each K8s component when we changes the value of flag. Currently, I am investigating a method which applies against all of component and does not need restart.

Reply:
@serathius 
In your strctured logging KEP, you are going to implement following structure.

> LoggingConfig structure should be implemented as part of k8s.io/component-base options and include a common set of flags for logger initialization.

Can we add `--request-id` parameter into this structure? Since `--request-id` parameter is also related to logging, I think it would be a simpler implementation to include it in this structure.

Reply:
@serathius 
I have additional questions.

set of klog flags is included in klog package. Does LoggingConfig structure is related to these klog flags?
if there is a component which does not use logger initialization via component-base, we need to implement this against such component?

Reply:
Ok, I think I understand the proposal now.  I'll defer to @serathius, but I don't think we want to introduce a new verbosity level for each piece of metadata we are considering adding to logs.

If we are very concerned about this adding to log size, we might instead prioritize adding this field to lower (>4) verbosity logs.  That way, we don't log this field very often by default, but turning up the verbosity adds many logs that contain the field.

Reply:
Thanks the comment.
I will wait reply from @serathius . I understand your opinion, because `--request-id` parameter's effect is similar to the existing verbosity level. (So, this may make the use of parameters difficult and complicated for users.)
 
If we use existing verbosity log(--v), I want to consider which level we should add request-id.